doc/developer/design/20231027_refresh_mvs.md

```diff
@@ -0,0 +1,276 @@
+
+### Implementation Alternative: Creating a New Dataflow at Each Refresh
+
+(The above alternatives are alternatives to `REFRESH EVERY` MVs, but this alternative is just about the implementation details of `REFRESH EVERY` MVs.)
+
+This approach would need similar replica management as the proposed solution, but instead of rehydrating an existing dataflow, we would always create a new dataflow to run the refresh. An advantage of this approach is that these dataflows could be single-time dataflows, and thus rely on one-shot SELECT optimizations. However, this approach would be more complicated, as we would have to separate the lifecycle of dataflows from the lifecycle of MVs. These new dataflows would be a kind of hybrid between Peek dataflows and MV dataflows, and would require new code in various parts of the Compute Controller. As we are under some time pressure to implement this feature, we opted for the simpler solution proposed in this document.
+
+### Implementation Alternative: Suspending the Replica Process to Disk
+
+As mentioned in the [scoping section](#out-of-scope), an alternative implementation that would cover also intermediate refresh intervals would be to suspend the replica process and save its entire state to S3 and a local disk cache, and restore it when a refresh is needed. [There are existing techniques for saving process state, e.g., CRIU.](https://faun.pub/kubernetes-checkpointing-a-definitive-guide-33dd1a0310f6) However, this would be a much bigger project than the proposed solution.
+
+## Automated Cluster Scheduling for `REFRESH EVERY` MVs
+
+TODO: updated discussion
+
+## Introspection / Observability
+
+For showing the `REFRESH` options of each materialized view, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refresh_options`, which will have a row for each non-trivial refresh option of each MV. The columns will be `(materialized_view_id text, interval interval, interval_aligned_to timestamptz, at timestamptz)`, and only those columns will have a non-null value that are relevant for the refresh option that the row is about (e.g., a `REFRESH EVERY` would be non-null in interval and interval_aligned_to). `REFRESH ON COMMIT` is considered a trivial refresh option, and won't show up in the table. (See Nikhil's [msg](https://materializeinc.slack.com/archives/C06535JL58R/p1701228087745239?thread_ts=1700595291.424189&cid=C06535JL58R) on an old design thread.)
```

Comment:
Is it possible for a single materialized view to have multiple columns in `mz_materialized_view_refresh_options`?

Reply:
Do you mean multiple rows? Yes, an MV can have multiple refresh options. For example, here is a MV with 3 refresh options: it refreshes at creation, every Thursday, and every Monday.
```
materialize=> 
create table t (x int);

CREATE CLUSTER c1 (SIZE '1', SCHEDULE = ON REFRESH (REHYDRATION TIME ESTIMATE = '10 sec'));

CREATE TABLE
CREATE CLUSTER
materialize=> 
CREATE MATERIALIZED VIEW mv3
IN CLUSTER c1
WITH (REFRESH = AT CREATION, REFRESH = EVERY '1 day' ALIGNED TO '2024-04-11 22:00:00', REFRESH = EVERY '1 day' ALIGNED TO '2024-04-15 22:00:00')
AS SELECT * FROM t;

CREATE MATERIALIZED VIEW
materialize=> 
SELECT name, mvro.*
FROM mz_internal.mz_materialized_view_refresh_options mvro, mz_catalog.mz_materialized_views mv
WHERE mv.id = mvro.materialized_view_id
ORDER BY mv.id;

 name | materialized_view_id | interval |  interval_aligned_to   |             at             
------+----------------------+----------+------------------------+----------------------------
 mv3  | u2                   |          |                        | 2024-04-11 15:33:13.256+00
 mv3  | u2                   | 24:00:00 | 2024-04-11 22:00:00+00 | 
 mv3  | u2                   | 24:00:00 | 2024-04-15 22:00:00+00 | 
(3 rows)
```